src/main/java/edu/kit/compiler/optimizations/unrolling/LoopVariableAnalysis.java

```diff
@@ -0,0 +1,232 @@
+            } else if (incLong == 0 || incLong < 0) {
+                return Optional.empty();
+            } else {
+                var difference = upperLong - lowerLong + incLong - 1;
+
+                return Optional.of(Long.divideUnsigned(difference, incLong));
+            }
+        }
+
+        private void assertCorrectNumSteps(long steps) {
+            var offset = step.asLong() * steps;
+            var result = new TargetValue(initial.asLong() + offset, initial.getMode());
+
+            assert steps >= 0;
+            assert relation.contains(initial.compare(bound)) || steps == 0;
+            assert !relation.contains(result.compare(bound));
+            assert steps == 0 || relation.contains(result.sub(step).compare(bound));
+        }
+    }
+
+    /**
+     * Recursively analyzes every path through a loop starting at the i-th
+     * predecessor of the loop variable. Returns a constant Tarval if the
+     * node is equal to the previous value of the loop variable plus a constant
+     * step (beware that the step may be zero).
+     */
+    private static TargetValue computeStepValue(Phi loopVariable, int i) {
+        loopVariable.getGraph().incVisited();
+        var zero = loopVariable.getMode().getNull();
+        return analyzeStepValue(loopVariable, zero, loopVariable.getPred(i));
+    }
+
+    private static TargetValue analyzeStepValue(Phi loopVariable,
+            TargetValue step, Node node) {
+        if (step.equals(BAD)) {
+            return BAD;
+        }
+
+        return switch (node.getOpCode()) {
+            case iro_Add -> {
+                var newStep = step.add(getConstValue(node.getPred(1)));
+                yield analyzeStepValue(loopVariable, newStep, node.getPred(0));
+            }
+            case iro_Phi -> {
+                if (node.equals(loopVariable)) {
+                    yield step;
+                } else {
+                    if (node.visited()) {
+                        yield BAD;
+                    } else {
+                        node.markVisited();
+                        var newStep = UNKNOWN;
+                        for (int i = 0; i < node.getPredCount() && !newStep.equals(BAD); ++i) {
+                            newStep = supremum(newStep, analyzeStepValue(
+                                    loopVariable, step, node.getPred(i)));
+                        }
+                        yield newStep;
+                    }
+                }
+            }
+            default -> BAD;
+        };
+    }
+
+    /**
+     * If the node is Const, returns its value. Otherwise returns BAD.
+     */
+    private static TargetValue getConstValue(Node node) {
+        return switch (node.getOpCode()) {
+            case iro_Const -> ((Const) node).getTarval();
+            default -> BAD;
+        };
+    }
+
+    /**
+     * Returns the supremum of the two Tarvals.
+     */
+    private static TargetValue supremum(TargetValue lhs, TargetValue rhs) {
```

Comment:
Just as a note: we also have the `TagetValueLatticeElement` class implementing this functionality.

Reply:
To be honest, for this very limited use case, I found the extra layer of abstraction a bit inconvenient.